---
title: "SQANTI3 OUTPUT COMPARISON REPORT"
author: "`r paste0('  ','Author', ': ',Sys.getenv('USERNAME'))`"
date: "`r paste0('  ','Date', ': ', format(Sys.time(), '%d %B %Y'))`"
output:
  html_document:
    number_sections: false
    toc: true
    toc_float: true
    toc_depth: 3
#    theme: spacelab
    css: style.css
    pandoc_args: [
      "+RTS", "-K64m",
      "-RTS"
    ]

---

<!-- Basic setting -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, cache = FALSE, echo = FALSE)
venn <- FALSE
```

<!-- Summary generation -->

## Summary table {.tabset .tabset-fade}

### SQANTI3 output

```{r}
t1.1
```

### Unique tag

```{r}
t1.2
```

<!-- Plot generation -->

## Gene characterization {.tabset .tabset-fade}

### Isoforms per gene

```{r}
p1.1
```

### Exon structure

```{r}
p1.2
```


## Structural category distribution

```{r}
p2
```

<!-- ADD ## Splice Junction Characterization -->

## TTS and TSS distances {.tabset .tabset-fade}

### Distance to TSS {.tabset .tabset-fade .tabset-dropdown}

#### FSM

```{r}
p4.1
```

#### ISM

```{r}
p4.2
```

### Distance to TTS {.tabset .tabset-fade .tabset-dropdown}

#### FSM

```{r}
p5.1
```

#### ISM

```{r}
p5.2
```

## CAGE peak distance {.tabset .tabset-fade .tabset-dropdown}

### FSM

```{r}
p6.1
```

### ISM

```{r}
p6.2
```

## Features of bad quality {.tabset .tabset-fade}

### RT-switching {.tabset .tabset-fade .tabset-dropdown}

#### FSM

```{r}
p7.1.1
```

#### NIC

```{r}
p7.1.2
```

#### NNC

```{r}
p7.1.3
```

<!-- ADD ## Features of good quality -->

## Complete comparison {.tabset .tabset-fade}

<!-- Print or not Venn Diagrams -->

```{r}
if (length(f_in) < 4) {
  venn<-TRUE
  asis_output("### Venn Diagrams\\n") # Header
  }
```

```{r, eval=venn}
grid.arrange(gTree(children = p9), top = "All isoforms")
```

### UpSet plot

```{r}
p10
grid.text("All isoforms",x = 0.65, y=0.95, gp=gpar(fontsize=20))
```

<!-- By structural category -->

## Comparison by structural category

<!-- Print or not Venn Diagrams -->

```{r, results='asis'}
if (length(f_in) < 4) {
  venn<-TRUE
  cat("\n\n###  Venn Diagrams {.tabset .tabset-pills}\n") # Header
  }
```

```{r, eval=venn, results='asis'}
for (i in 1:length(p11)){
  cat("\n\n#### ", str_cat[i],"\n")
  grid.arrange(gTree(children = p11[[i]]), top = str_cat[i])
}
```

### UpSet plot {.tabset .tabset-fade}

```{r, results='asis'}
for (i in 1:length(p12)){
  cat("\n\n#### ", str_cat[i],"\n")
  print(p12[i])
  grid.text(str_cat[i],x = 0.65, y=0.95, gp=gpar(fontsize=20))
}
```

<!-- Print or not TSS and TSS coords stats -->

```{r,results='asis'}
if (TSS_TTS_coord == TRUE) {
  cat("## Standard deviation{.tabset .tabset-fade}\n\n") # Header
  cat("### UJC metrics {.tabset .tabset-fade .tabset-dropdown}\n\n")
  cat("#### TSS\n\n")
  print(p15.1)
  cat("\n\n")
  cat("#### TTS\n\n")
  print(p15.2)
  cat("\n\n")
  cat("### Standard deviation for each pipeline {.tabset .tabset-fade .tabset-dropdown}\n\n")
  cat("#### TSS\n\n")
  print(p13)
  cat("\n\n")
  cat("#### TTS\n\n")
  print(p14)
  cat("\n\n")
}

```

## Gene metrics{.tabset .tabset-fade}

### Raw

```{r}
p17.1
p17.2
p17.3
p17.4
```

### Log 2

```{r}
p17.5
p17.6
p17.7
```


```{r, results='asis'}
if (TSS_TTS_coord == TRUE){
  cat("\n\n## Isoform metrics{.tabset .tabset-fade}\n\n")
  cat("\n\n### TSS{.tabset .tabset-fade}\n\n")
  cat("\n\n#### Raw\n\n")
  print(p16.1)
  print(p16.2)
  print(p16.3)
  print(p16.4)
  cat("\n\n#### Log 2\n\n")
  print(p16.9)
  print(p16.10)
  print(p16.11)
  print(p16.12)
  cat("\n\n### TTS{.tabset .tabset-fade}\n\n")
  cat("\n\n#### Raw\n\n")
  print(p16.5)
  print(p16.6)
  print(p16.7)
  print(p16.8)
  cat("\n\n#### Log 2\n\n")
  print(p16.13)
  print(p16.14)
  print(p16.15)
  print(p16.16)
}
```

## LRGASP metrics {.tabset .tabset-fade}

### FSM evaluation

```{r}
if (lrgasp == TRUE){
  t4
}
```

### ISM evaluation

```{r}
if (lrgasp == TRUE){
  t5
}
```

### NIC evaluation

```{r}
if (lrgasp == TRUE){
  t6
}
```

### NNC evaluation

```{r}
if (lrgasp == TRUE){
  t7
}
```

### SIRVs evaluation

```{r}
if (lrgasp == TRUE){
  t3
}
```


## Presence/Ausence table

```{r}
t2
```

## SessionInfo

```{r}
sessionInfo()
```
