## Table of contents:

* <a href="#output">SQANTI3 QC output</a>

    * <a href="#main">Main output</a>

    * <a href="#transcriptome">Transcriptome files</a>

    * <a href="#report">Report</a>

    * <a href="#supplementary">Supplementary output</a>

* <a href="#classifcols">Glossary of classification file columns </a>

* <a href="#junctioncols">Glossary of junction file columns </a>

__________________________________

<a name="output"></a>

## SQANTI3 QC output

As a result of [running SQANTI3 QC](Running-SQANTI3-Quality-Control.md), the tool will create a series of output files in the specified directory (`--dir` or `-d` flag in the QC script), all of which will be named using the supplied prefix (`--output` or `-o` flag in the QC script). Given that SQANTI3 also runs several other tools internally (e.g. STAR, kallisto, TD2...), a series of subfolders will also appear in the output directory. 

The [example/SQANTI3_QC_output](https://github.com/ConesaLab/SQANTI3/tree/master/example/SQANTI3_QC_output) subfolder in the SQANTI3 main directory contains an updated example of what the output of SQ3 QC looks like:

![SQANTI3 QC Output Example](https://raw.githubusercontent.com/aarzalluz/figures_public/master/SQANTI3/example_output_SQ3-QC.png)

This is the result of [running SQANTI3 QC on example data](Tutorial:-running-SQANTI3-on-an-example-dataset.md), i.e. a subset of the UHR dataset. A more detailed explanation of the different output files included below.

<a name="main"></a>

### Main output: classification and junctions files

SQANTI3 characterizes the input transcriptome by computing a series of attributes by transcript, which are written to the classification file, and a series of attributes by junction, which are written to the junctions file.

- **Classification file**: a tab-separated file where transcripts are rows and QC attributes computed by SQANTI3 are columns. Isoforms are identified by their ID in the input long-read GTF file. For a full glossary of columns and their meaning, [see below](#classifcols).

- **Junctions file**: a tab-separated file containing each junction for every PB isoform as rows. Rows are identified by the isoform ID and the junction number (shown in order, i.e. the 1st, 2nd, 3rd... ith junction), which are included in the two first columns. Take into account that the same junction may appear multiple times if they are shared by multiple PB isoforms, for instance, when several FSM and ISM transcripts associated to the same reference isoform are found. For a full glossary of columns and their meaning, [see below](#glossary-of-junctions-file-columns-junctionstxt)


<a name="transcriptome"></a>

### Transcriptome files

- **GTF file**: the `_corrected.gtf` file constitutes the **main transcriptome output of SQANTI3**. It contains the annotation for all transcripts supplied to SQANTI3. This file may be used for downstream analyses such as quantification, however, we recommend filtering your transcriptome using the QC attributes in SQANTI3 before proceeding with your analyses. For this purpose, you may apply one of the [two filters available in SQANTI3](Running-SQANTI3-filter.md) or, alternatively, define your own filtering criteria.

- **FASTA file**: the `_corrected.fasta` file contains the genomic sequences of transcripts supplied to SQANTI3 QC, obtained using the `_corrected.gtf` file and the provided reference genome.

- **FAA file**: the `_corrected.faa` file contains FASTA-formatted protein sequences derived from transcript sequences. These sequences are obtained by running the **TransDecoder2 (TD2) ORF prediction** algorithm and will only be output if the `--include_ORF` flag is supplied.

- **CDS GFF file**: the `_corrected.cds.gff3` file is generated by expanding the GTF file to include **coding sequence (CDS)** information (obtained from TD2 ORF predictions). This means that, in addition to exons and transcripts, CDS will also be included as a feature in the 3rd GFF column.

- **tappAS GFF3 file** (optional): if the `--isoAnnotLite` argument is used, **IsoAnnotLite** will be run internally and a tappAS-formatted GFF3 file containing all isoforms in the input transcriptome will be generated in the output folder. This allows direct usage of a SQANTI3-processed transcriptome in tappAS. Of note, if a pre-annotated tappAS GFF3 file is supplied to SQANTI3 QC via the `--gff3` flag, the tappAS GFF3 annotation will also contain functional feature transferred by position (see details on how IsoAnnotLite works [here](IsoAnnotLite.md)).

#### Note on indel correction and GTF/FASTA output files

SQ3 currently accepts both GTF (default) and FASTA transcriptome files as input. If a **FASTA input file** is supplied, sequences are mapped to the genome to generate the `_corrected.gtf` file. Using this GTF and the reference genome, SQ3 performs **indel correction** and generates the `_corrected.fasta` file. Please note that this type of reference-based correction **will remove SNPs**, therefore, users who are interested in this kind of information should refrain from using a FASTA file as input. Alternatively, if users supply a **GTF input file**, SQ3 will NOT perform indel correction, just extract transcript sequences using the reference genome and the transcriptome GTF file, which will be equivalent to the `_corrected.gtf` file.


<a name="report"></a>

### SQANTI QC report

The SQANTI QC report is generated by the [SQANTI3_report.R](https://github.com/ConesaLab/SQANTI3/blob/tree/master/utilities/SQANTI3_report.R) script, which takes the **classification** (e.g. `UHR_chr22_classification.txt`) and **junctions** (e.g. `UHR_chr22_junctions.txt`) files. 

The report includes over 150 plots, therefore, in order to ease interpretation, we recently implemented the `--report` argument with four options: `pdf`, `html`, `both` and `skip`. This means that the report can be created as an **interactive HTML document** as well as (or alternatively to) a **PDF report** -or skipped entirely if the user does not wish to see a report for the QC run.

Users who are proficient in R are welcome to modify the R script to add new figures -we will be constantly adding new figures as well.


<a name="supplementary"></a>

### Supplementary output

- **Parameter file** (`_params.txt`): plain text file including the full paths of all files supplied to SQANTI3, as well as any additional parameters included in the call to `sqanti3_qc.py`. Here is an example of the fields included in the parameter file for a full SQ3 QC run starting from raw short-read data.

- **STAR folders** contain the different outputs of the [STAR aligner](https://github.com/alexdobin/STAR), which is [run internally using provided short-reads](Running-SQANTI3-Quality-Control.md#providing-short-reads) during QC.

  - **STAR_index**: STAR index for the supplied reference genome. If you wish to supply a pre-computed index to SQ3 QC, see the shortcut described in [issue #130](https://github.com/ConesaLab/SQANTI3/issues/130).
  - **STAR_mapping**: STAR mapping files. SQANTI3 uses short-read coverage information (`_SJ.out.tab` file) and mapping BAM files (`_Aligned.sortedByCoord.bam`) to compute short-read coverage, isoform expression (see kallisto output below) and the TSS ratio metric.

- **Kallisto output folder**: this folder will contain the results of running [kallisto](https://pachterlab.github.io/kallisto/) to compute isoform expression estimates using the different short-read replicates supplied. In it, users will find one subfolder per replicate, in which kallisto abundance files (i.e. quantification matrices) are stored.

- **RT-switching (RTS) folder**: output files generated by the RT-switching prediction script (`utilities/rt_switching.py`).

- **GMST folder**: output files generated after running GeneMarkST (`utilities/gmst`) for ORF prediction.

- **genePred files**: reference annotation and long-read transcriptome GTF files are converted to genePred format during isoform classification. The two `.genePred` files found in the output folder constitute the intermediate files that are generated in this process.

<a name="classifcols"></a>

## Glossary of classification file columns (`classification.txt`)

The output `_classification.txt` has the following fields:

1. `isoform`: the isoform ID. Usually in `PB.X.Y` format.
2. `chrom`: chromosome.
3. `start`: isoform start position.
4. `end`: isoform end position.
5. `strand`: strand.
6. `length`: isoform length.
7. `exons`: number of exons.
8. `structural_category`: one of the categories ["full-splice_match", "incomplete-splice_match", "novel_in_catalog", "novel_not_in_catalog", "genic", "antisense", "fusion", "intergenic", "genic_intron"]
9. `subcategory`: one of the possible subcategories for each category (see [isoform categories and subcategories](SQANTI3-isoform-classification:-categories-and-subcategories.md))
10. `FSM_class`: This feature classifies the transcript according to the expression of other isoforms in the gene to which the transcript belongs. Transcripts belonging to genes that only express one isoform are classified as A. Transcripts belonging to genes that express more than one isoform but none is a FSM are classified as B. Transcripts belonging to genes which express more than one isoform and other isoforms and at least one is a FSM are classified as C.
11. `associated_gene`: the reference gene name.
12. `associated_transcript`: the reference transcript name.
13. `ref_length`: matched reference transcript length. For transcripts in the `novel_in_catalog` and `novel_not_in_catalog` categories, the best matching reference transcript is chosen based on the highest number of shared splice junctions and the smallest difference in exon count.
14. `ref_exons`: reference transcript number of exons.
15. `diff_to_TSS`: distance of query isoform 5' start to reference transcript start end. Negative value means query starts downstream of the reference TSS, so the transcript is shorter than reference transcript.
16. `diff_to_TTS`: distance of query isoform 3' end to reference annotated end site. Negative value means query ends upstream of TTS, so the transcript is shorter than reference transcript.
17. `diff_to_gene_TSS`: distance of query isoform 5' start to the closest start end of *any* transcripts of the matching gene. This field is different from `diff_to_TSS` since it's looking at all annotated starts of a gene. Negative value means query starts downstream of TSS.
18. `diff_to_gene_TTS`: distance of query isoform 3' end to the closest end of *any* transcripts of the matching gene. Negative value means query ends upstream of TTS. 
19. `RTS_stage`: TRUE if one of the junctions could be a RT switching artifact.
20. `all_canonical`: TRUE if all junctions have canonical splice sites.
21. `min_sample_cov`: sample with minimum coverage.
22. `min_cov`: minimum junction coverage based on short read STAR junction output file. NA if no short read given.
23. `min_cov_pos`: the junction that had the fewest coverage. NA if no short read data given.
24. `sd_cov`: standard deviation of junction coverage counts from short read data. NA if no short read data given.
25. `FL` or `FL.<sample>`: FL count associated with this isoform per sample if `--fl_count` is provided, otherwise NA.
26. `n_indels`: total number of indels based on alignment.
27. `n_indels_junc`: number of junctions in this isoform that have alignment indels near the junction site (indicating potentially unreliable junctions).
28. `bite`: TRUE if contains at least one "bite" positive SJ.
29. `iso_exp`: short read expression for this isoform if `--expression` is provided, otherwise NA.
30. `gene_exp`: short read expression for the gene associated with this isoform (summing over all isoforms) if `--expression` is provided, otherwise NA.
31. `ratio_exp`: ratio of `iso_exp` to `gene_exp` if `--expression` is provided, otherwise NA.
32. `coding`: Coding potential capacity according to GeneMarkS-T. It can take values of "coding" or "non_coding"
33. `ORF_length`: predicted ORF length (including introns) in nucleotides.
34. `CDS_length`: predicted CDS length in nucleotides.
35. `protein_length`: predicted protein length in aminoacids. 
36. `CDS_start`: CDS start.
37. `CDS_end`: CDS end.
38. `CDS_genomic_start`: genomic coordinate of the CDS start. If on - strand, this coord will be greater than the end.
39. `CDS_genomic_end`: genomic coordinate of the CDS end. If on - strand, this coord will be smaller than the start.
40. `psauron_score`: Value between 0 and 1 indicating the coding potential of the transcript according to the psauron algorithm. Higher values indicate higher coding potential.
41. `CDS_type`: one of the following based on the presence/absence of start and stop codons: "complete", "5prime_partial", "3prime_partial", "internal"
42. `predicted_NMD`: TRUE if there's a predicted ORF and CDS ends at least 50bp before the last junction; FALSE if otherwise. NA if non-coding.
43. `perc_A_downstream_TTS`: percent of genomic "A"s in the downstream 20 bp window. If this number is high (say > 0.8), the 3' end site of this isoform is probably not reliable.
44. `seq_A_downstream_TTS`: sequence of the downstream 20 bp window.
45. `dist_to_CAGE_peak`: distance to closest TSS based on CAGE Peak data. Negative means the middle of the cage peak is upstream of the isoform TSS and positive means downstream of the isoform TSS. Strand-specific. SQANTI3 only searches for nearby CAGE Peaks within 10000 bp of the PacBio transcript start site. Will be `NA` if none are found within 10000 bp.
46. `within_CAGE_peak`: TRUE if the transcript start site is within a CAGE Peak. 
47. `dist_to_polyA_site`: distance to the closest polyA site, based on polyA site data (e.g. Quant-seq).
48. `within_polyA_site`: TRUE if the transcript start site is within a polyA site, retrieved from polyA site data (e.g. Quant-seq).
49. `polyA_motif`: if `--polyA_motif_list` is given, shows the top ranking polyA motif found within 50 bp upstream of end.
50. `polyA_dist`: if `--polyA_motif_list` is given, shows the location of the  last base of the hexamer. Position 0 is the putative poly(A) site. This distance is hence always negative because it is upstream. 
51. `polyA_motif_found`: TRUE if a polyA motif given via `--polyA_motif_list` is detected in the 3'end of the transcript.  
52. `ratio_TSS`: Using Short-Read data, we measure the mean coverage of the 100bp upstream and downstream a reported TSS. Then we calculate the ratio *coverage inside isoform + 0.01/ coverage outside isoform + 0.01*. If several SR samples are provided, `ratio_TSS` will represent the maximum value of the ratios across the samples. This means that if an isoform has a `ratio_TSS` greater than 1 it is more likely that its TSS is true. Meanwhile, if the `ratio_TSS` is close or lower than 1, the SR coverage is similar inside and outside the isoform, something that we wouldn't expect if the TSS was true.


<a name="junctioncols"></a>

## Glossary of junctions file columns (`junctions.txt`)

The `_junctions.txt` file contains the following columns:

1. `isoform`: Isoform ID.
2. `junction_number`: The i-th junction of the isoform.
3. `chrom`: Chromosome.
4. `strand`: Strand.
5. `genomic_start_coord`: Start of the junction (1-based), note that if on - strand, this would be the junction acceptor site instead.
6. `genomic_end_coord`: End of the junction (1-based), note that if on - strand, this would be the junction donor site instead.
7. `transcript_coord`: Currently not implemented. Ignore.
8. `junction_category`: `known` if the (donor-acceptor) combination is annotated in the GTF file, `novel` otherwise. Note that it is possible to have a `novel` junction even though both the donor and acceptor site are known, since the combination might be novel.
9. `start_site_category`: `known` if the junction start site is annotated. If on - strand, this is actually the donor site.
10. `end_site_category`: `known` if the junction end site is annotated. If on - strand, this is actually the acceptor site.
11. `diff_to_Ref_start_site`: distance to closest annotated junction start site. If on - strand, this is actually the donor site.
12. `diff_to_Ref_end_site`: distance to closest annotated junction end site. If on - strand, this is actually the acceptor site.
13. `bite_junction`: Applies only to novel splice junctions. If the novel intron partially overlaps annotated exons the bite value is TRUE, otherwise it is FALSE.
14. `splice_site`: Splice motif.
15. `RTS_junction`: TRUE if junction is predicted to be a template switching artifact.
16. `indel_near_junct`: TRUE if there is alignment indel error near the junction site, indicating potential junction incorrectness.
17. `sample_with_cov`: If `--coverage` (short read junction coverage info) is provided, shows the number of samples (cov files) that have uniquely mapped short reads that support this junction.
18. `total_coverage_unique/multi`: Total number of uniquely or multi-mapped short read support from all samples that cover this junction.

